[AA-1300] Admin app 2.2 Azure testing

CSR2017 opened the pull request:
@vimayya @stephenfuqua @plioi
Please help review and merge the changes.
Changes include separate powershell script file and ARM templates which are specifically defined for AdminApp 2.2.0 in-place upgrade in Azure environment.
This implementation designed to work with existing azure resource group and ODS API. 
Added new folder and files, since don't want to customize the existing code for unique purpose/ use case.

Thanks

CSR2017 committed:
Adding just Admin app deployment ARM template and script file

CSR2017 committed:
Set connection strings

CSR2017 committed:
Move files to separate location. Update powershell script to support db migration

CSR2017 committed:
git ignore sql script files

CSR2017 committed:
Package name update and include encryption key

stephenfuqua commented on Application/Published/Upgrade-AdminApp/Deploy-EdFiOdsAdminApp.ps1:
```diff
@@ -0,0 +1,220 @@
+	[Parameter(Mandatory=$true)] 
+	$ResourceGroupName,
+
+	[ValidateSet("East US", "South Central US")]
```
Why is this restricted to only these two resource group locations?

CSR2017 replied:
@stephenfuqua 
These are the existing supported resource group location.

stephenfuqua replied:
What a strange restriction. I will ask Jason if it can be removed.

stephenfuqua replied:
never mind, I'm just going to ignore it for now.

stephenfuqua commented on Application/Published/Upgrade-AdminApp/Deploy-EdFiOdsAdminApp.ps1:
```diff
@@ -0,0 +1,220 @@
+
+function Get-SqlServerInfo()
+{	
+	$hostName = Read-Host -Prompt "SQL Server Hostname (ex: sql.mydomain.com,1433)"
```
Why not provide the host name and port as function parameters? If continuing with the user prompt, would it make sense to specify "SQL Server Hostname and Port Number (ex: sql.mydomain.com,1433)"

CSR2017 replied:
@stephenfuqua 
I tested with just the hostname. It is working fine. Will be removing the port name. Since the default port is used down the line.

stephenfuqua commented on Application/Published/Upgrade-AdminApp/Deploy-EdFiOdsAdminApp.ps1:
```diff
@@ -0,0 +1,220 @@
+{	
+	$hostName = Read-Host -Prompt "SQL Server Hostname (ex: sql.mydomain.com,1433)"
+	$hostName = $hostName.Replace(':', ',')
+	$adminCredentialsPrompt = "Please enter a username and password for your SQL Server.  These credentials will be used to create new database users for your Ed-Fi ODS installation."
```
Same question here - why not provide as function parameters?

If this is an existing pattern then I'm not trying to change it right now. But I don't know the existing pattern, and I'm just reading this script from top to bottom. Seems strange to have a mix of items passed in as function parameters and those that are prompted from the user at the command line.

stephenfuqua commented on Application/Published/Upgrade-AdminApp/Deploy-EdFiOdsAdminApp.ps1:
```diff
@@ -0,0 +1,220 @@
+.PARAMETER ProductionApiUrl
+	Existing ODS/ API Url that Admin App can connect
+.PARAMETER EncryptionKey
+	Base64-encoded 256 bit key appropriate for use with AES encryption
```
Should there be a comment that this is optional, and a key will be created if one is not provided?

stephenfuqua commented on Application/Published/Upgrade-AdminApp/Deploy-EdFiOdsAdminApp.ps1:
```diff
@@ -0,0 +1,220 @@
+
+function IsDBInstalled([string]$Server, [string]$database)
+{	 
+	$t=Invoke-Sqlcmd -ServerInstance $Server -Username  $user -Password  $pwd -Database "master" -Query "select 1 from sys.databases where name='$database'" -OutputSqlErrors $true 
```
This is a very long line, what do you think about refactoring to an argument array?

stephenfuqua commented on Application/Published/Upgrade-AdminApp/Deploy-EdFiOdsAdminApp.ps1:
```diff
@@ -0,0 +1,220 @@
+{
+	$destinationZipFile = "$PSScriptRoot/Database.zip"
+	$scriptFilesPath = "$PSScriptRoot/ScriptFiles"
+	Invoke-WebRequest -Uri "https://odsassets.blob.core.windows.net/public/adminapp/Release/2.2.0/Edfi.suite3.ods.adminapp.database.zip" -OutFile $destinationZipFile
```
What do you think about moving the `Uri` value to a constant at the top of the file? That will make this line shorter / easier to read, and it will be easier to find and update this value for the next release.

Also, should it use `$AdminAppVersion` instead of hard-coding `2.2.0`?

stephenfuqua commented on Application/Published/Upgrade-AdminApp/Deploy-EdFiOdsAdminApp.ps1:
```diff
@@ -0,0 +1,220 @@
+	foreach ($s in $scripts)
+	{   
+		Write-Host "Running Script : " $s.Name -ForegroundColor Yellow
+		$tables=Invoke-Sqlcmd -ServerInstance $Server -Username  $user -Password  $pwd -Database  $database -InputFile $s.FullName -ErrorAction 'Stop' -querytimeout (65535)
```
Can this be refactored to use an argument array?

stephenfuqua commented on Application/Published/Upgrade-AdminApp/Deploy-EdFiOdsAdminApp.ps1:
```diff
@@ -0,0 +1,220 @@
+	$Server =  $sqlServerInfo.HostName
+	$database = "EdFi_Admin"
+	$user = $sqlServerInfo.AdminCredentials.UserName
+	$pwd = (New-Object PSCredential $user, $sqlServerInfo.AdminCredentials.Password).GetNetworkCredential().Password	
```
How come you can't just use `$sqlServerInfo.AdminCredentials.Password` here?

CSR2017 replied:
@stephenfuqua 
The value is secure string. I want to have plain text for this use case.

stephenfuqua commented on Application/Published/Upgrade-AdminApp/Deploy-EdFiOdsAdminApp.ps1:
```diff
@@ -0,0 +1,220 @@
+	$user = $sqlServerInfo.AdminCredentials.UserName
+	$pwd = (New-Object PSCredential $user, $sqlServerInfo.AdminCredentials.Password).GetNetworkCredential().Password	
+	 
+	IsDBInstalled $Server $database	 
```
Should `$user` and `$pwd` be arguments to this function call?

CSR2017 committed:
Included SQL host and credentials as parameters. Refactored methods to take required parameters.

stephenfuqua commented on Application/Published/Upgrade-AdminApp/Deploy-EdFiOdsAdminApp.ps1:
```diff
@@ -0,0 +1,220 @@
+		$EncryptionKey = [System.Convert]::ToBase64String($aes.Key)
+	}
+	
+	$deployParameters = New-Object -TypeName Hashtable
+	$deployParameters.Add("version", $AdminAppVersion)
+	$deployParameters.Add("edition", $Edition)
+	$deployParameters.Add("appInsightsLocation", $AppInsightLocation)
+	$deployParameters.Add("odsInstanceName", $ResourceGroupName)
+	$deployParameters.Add("sqlServerAdminLogin", $SQLServerUserName)
+	$deployParameters.Add("sqlServerAdminPassword", $SQLServerPassword)
+	$deployParameters.Add("productionApiUrl", $ProductionApiUrl)	
+	$deployParameters.Add("metadataCacheTimeOut", $CacheTimeOut)
+	$deployParameters.Add("adminAppNameToDeploy", $AdminAppName)
+	$deployParameters.Add("encryptionKey", $EncryptionKey)
```
You don't need to change anything... just commenting for future reference. Typically I see this written in a simpler format:
```
	$deployParameters = @{
	    version = $AdminAppVersion
	    edition = $Edition
	    appInsightsLocation = $AppInsightLocation
            ... etc
        }
```

stephenfuqua commented on Application/Published/Upgrade-AdminApp/Deploy-EdFiOdsAdminApp.ps1:
```diff
@@ -0,0 +1,220 @@
+	Try
+	{
+		Write-Host "Deploying ODS Admin App"		
+		$deploymentResult = New-AzureRmResourceGroupDeployment -DeploymentDebugLogLevel All -Name ((Get-ChildItem $templateFile).BaseName + '-' + ((Get-Date).ToUniversalTime()).ToString('MMdd-HHmm')) `
+									   -ResourceGroupName $ResourceGroupName `
+									   -TemplateFile $templateFile `
+									   -TemplateParameterFile $templateParametersFile `
+									   @deployParameters `
+									   -Force -Verbose -ErrorAction Stop
```
Here's a possible formatting improvement. However, there's actually a better approach: why not move these other parameters into $deployParameters so that they're all defined in one place? For `-Force` and `-Verbose`, use `Force = $True` and `Verbose = $True` when defining the [splat](https://devblogs.microsoft.com/scripting/use-splatting-to-simplify-your-powershell-scripts/).

```suggestion
                $name = ((Get-ChildItem $templateFile).BaseName + '-' + ((Get-Date).ToUniversalTime()).ToString('MMdd-HHmm'))
		$deploymentResult = New-AzureRmResourceGroupDeployment `
                                  -DeploymentDebugLogLevel All `
                                  -Name $name `
                                  -ResourceGroupName $ResourceGroupName `
                                  -TemplateFile $templateFile `
                                  -TemplateParameterFile $templateParametersFile `
                                  @deployParameters `
                                  -Force `
                                  -Verbose `
                                  -ErrorAction Stop
```

stephenfuqua commented on Application/Published/Upgrade-AdminApp/Deploy-EdFiOdsAdminApp.ps1:
```diff
@@ -0,0 +1,220 @@
+.PARAMETER SQLServerPassword
+	Password for your SQL Server.
+.PARAMETER EncryptionKey
+	Base64-encoded 256 bit key appropriate for use with AES encryption. This is optional parameter. A key will be created if one is not provided.
```
```suggestion
	Base64-encoded 256 bit key appropriate for use with AES encryption. This is 
	an optional parameter. A key will be created if one is not provided.
```

stephenfuqua commented on Application/Published/Upgrade-AdminApp/Deploy-EdFiOdsAdminApp.ps1:
```diff
@@ -0,0 +1,220 @@
+.PARAMETER ResourceGroupName
+	Existing resource group name.
+.PARAMETER AdminAppName
+    A friendly name to help identify AdminApp within Azure.  Must be no more than 64 characters long.
```
```suggestion
    A friendly name to help identify AdminApp within Azure.  Must be no more 
    than 64 characters long.
```

stephenfuqua commented on Application/Published/Upgrade-AdminApp/Deploy-EdFiOdsAdminApp.ps1:
```diff
@@ -0,0 +1,220 @@
+.PARAMETER EncryptionKey
+	Base64-encoded 256 bit key appropriate for use with AES encryption. This is optional parameter. A key will be created if one is not provided.
+.PARAMETER TemplateFileDirectory
+	Points the script to the directory that holds the Ed-Fi ODS install templates. By default that directory is the same as the one that contains this script.
```
```suggestion
	Points the script to the directory that holds the Ed-Fi ODS install templates.
	By default that directory is the same as the one that contains this script.
```

stephenfuqua commented on Application/Published/Upgrade-AdminApp/Deploy-EdFiOdsAdminApp.ps1:
```diff
@@ -0,0 +1,220 @@
+.PARAMETER Edition
+	Edition (Test, Release) of the Azure Deploy scripts to deploy.  Defaults to Release.
+.EXAMPLE
+	.\Deploy-EdFiOds.ps1 -ResourceGroupName "Ed-Fi-Ods-Resourcegroup" -AdminAppName "AdminApp-Latest" -AppInsightLocation "South Central US" -ProductionApiUrl "https://edfiodsapiwebsite-production-yuw8iui32.azurewebsites.net"
+    Deploys the provided version of the AdminApp to the South Central US Azure region with the default instance name
```
```suggestion
	@params = @{
        ResourceGroupName = "Ed-Fi-Ods-Resourcegroup"
        AdminAppName = "AdminApp-Latest"
        AppInsightLocation = "South Central US"
        ProductionApiUrl = "https://edfiodsapiwebsite-production-yuw8iui32.azurewebsites.net"
    }
    .\Deploy-EdFiOds.ps1  @params
	
    Deploys the provided version of the AdminApp to the South Central US Azure 
    region with the default instance name
```

stephenfuqua commented on Application/Published/Upgrade-AdminApp/Deploy-EdFiOdsAdminApp.ps1:
```diff
@@ -0,0 +1,220 @@
+	Expand-Archive $destinationZipFile -DestinationPath $scriptFilesPath -Force	
+	
+	$database = "EdFi_Admin"
+	$pwd = (New-Object PSCredential $SQLServerUserName, $SQLServerPassword).GetNetworkCredential().Password	
```
`$SQLServerPassword` is already a plain string, so I don't think you need this line anymore.

CSR2017 replied:
@stephenfuqua 
We are receiving as Secure string. So need this line to convert.

stephenfuqua replied:
You're right, I misread the function argument.

stephenfuqua commented on Application/Published/Upgrade-AdminApp/Deploy-EdFiOdsAdminApp.ps1:
```diff
@@ -0,0 +1,220 @@
+			InputFile = $s.FullName			
+		}
+
+		$tables = Invoke-Sqlcmd @arguments -ErrorAction 'Stop' -querytimeout (65535)
```
I believe that's an 18 hour timeout 😆 .

CSR2017 committed:
Alignment changes.

Description alignment

stephenfuqua reviewed (commented):
Looks nice now!

vimayya commented:
@CSR2017 Is the intension here is to keep these additional scripts until next release and discard it when default release supports admin app 2.2 or higher or this needs to be kept longer of older version upgrades?

CSR2017 committed:
Automate Admin app specific tables and schema deletion

Update message with correct value
